bigquery/google/cloud/bigquery/dbapi/cursor.py

```diff
@@ -0,0 +1,280 @@
+            return operation
+
+        if isinstance(parameters, collections.Mapping):
+            return self._format_operation_dict(operation, parameters)
+
+        return self._format_operation_list(operation, parameters)
+
+    def execute(self, operation, parameters=None):
+        """Prepare and execute a database operation.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Mapping[str, Any] or Sequence[Any]
+        :param parameters: Optional dictionary or sequence of parameter values.
+        """
+        self._query_results = None
+        self._page_token = None
+        self._has_fetched_all_rows = False
+        client = self.connection._client
+        job_id = str(uuid.uuid4())
+
+        # The DB-API uses the pyformat formatting, since the way BigQuery does
+        # query parameters was not one of the standard options. Convert both
+        # the query and the parameters to the format expected by the client
+        # libraries.
+        formatted_operation = self._format_operation(
+            operation, parameters=parameters)
+        query_parameters = _helpers.to_query_parameters(parameters)
+
+        query_job = client.run_async_query(
+            job_id,
+            formatted_operation,
+            query_parameters=query_parameters)
+        query_job.use_legacy_sql = False
+        query_job.begin()
+        _helpers.wait_for_job(query_job)
+        query_results = query_job.results()
+
+        # Force the iterator to run because the query_results doesn't
+        # have the total_rows populated. See:
+        # https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3506
+        query_iterator = query_results.fetch_data()
+        try:
+            six.next(iter(query_iterator))
+        except StopIteration:
+            pass
+
+        self._query_data = iter(
+            query_results.fetch_data(max_results=self.arraysize))
+        self._set_rowcount(query_results)
+        self._set_description(query_results.schema)
+
+    def executemany(self, operation, seq_of_parameters):
+        """Prepare and execute a database operation multiple times.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Sequence[Mapping[str, Any] or Sequence[Any]]
+        :param parameters: Sequence of many sets of parameter values.
+        """
+        for parameters in seq_of_parameters:
+            self.execute(operation, parameters)
+
+    def fetchone(self):
+        """Fetch a single row from the results of the last ``execute*()`` call.
+
+        :rtype: tuple
+        :returns:
+            A tuple representing a row or ``None`` if no more data is
+            available.
+        """
+        if self._query_data is None:
+            raise exceptions.InterfaceError(
+                'No query results: execute() must be called before fetch.')
+
+        try:
+            return six.next(self._query_data)
```

Comment:
So this might be a fetch, or it might be just grabbing an item from a page (though the default page size is 1?)